File: bench/langs/Ledger.scala
package bench

/** Keeps the running balance for one account. */
class Ledger(owner: String) {

  /** Bills the account and returns what is left. */
  def charge(amount: Int): Int = {
    if (amount > Ledger.Limit) {
      refuse(amount)
    } else {
      commit(amount)
    }
  }

  private def refuse(amount: Int): Int = {
    warn(owner)
    0
  }

  private def commit(amount: Int): Int = {
    write(owner, amount)
    amount
  }
}

object Ledger {
  val Limit = 5000
}

File: bench/langs/ledger.cpp
// Keeps the running balance for one account.
#include "ledger.hpp"

namespace bench {

const int Limit = 5000;

class Ledger {
public:
    explicit Ledger(std::string owner) : owner_(std::move(owner)) {}

    // Bills the account and returns what is left.
    int charge(int amount) {
        if (amount > Limit) {
            return refuse(amount);
        }
        return commit(amount);
    }

private:
    int refuse(int amount) {
        warn(owner_);
        return 0;
    }

    int commit(int amount) {
        write(owner_, amount);
        return amount;
    }

    std::string owner_;
};

}
